Tools/autotest/common.py

```diff
@@ -1467,117 +1467,131 @@ def wait_seconds(self, seconds_to_wait):
-            m = self.mav.recv_match(type='ATTITUDE', blocking=True)
-            p = math.degrees(m.pitch)
-            r = math.degrees(m.roll)
-            self.progress("Pitch %d Roll %d" % (p, r))
-            if math.fabs(p - pitch) <= accuracy:
-                self.progress("Attained pitch %d" % pitch)
-                return True
-        raise WaitPitchTimeout("Failed to attain pitch %d" % pitch)
+            if called_function is not None:
+                called_function(last_value, target)
+            if target_retreiver is not None:
+                last_value = target_retreiver()
+            if self.get_sim_time_cached() - last_print_time > 1:
+                self.progress("%s=%0.2f (want %f +- %f)" %
+                              (value_name, last_value, target, accuracy))
+                last_print_time = self.get_sim_time_cached()
+            if validator is not None:
+                is_value_valid = validator(last_value, target)
+            else:
+                is_value_valid = math.fabs(last_value - target) <= accuracy
+            if is_value_valid:
+                mean_value += last_value
+                counter += 1.0
+                if duration_start is None:
+                    duration_start = self.get_sim_time_cached()
+                if self.get_sim_time_cached() - duration_start >= maintain_target_time:
+                    self.progress("Attained %s=%f" % (value_name, mean_value / counter))
+                    return True
+            else:
+                duration_start = None
+                mean_value = 0.0
+                counter = 0.0
+        raise AutoTestTimeoutException("Failed to attain %s %f, reach %f" % (value_name, target, (mean_value / counter) if counter != 0.0 else last_value))
 
-    def wait_heading(self, heading, accuracy=5, timeout=30):
+    def wait_heading(self, heading, accuracy=5, maintain_target_time=0, timeout=30, called_function=None):
         """Wait for a given heading."""
-        tstart = self.get_sim_time()
-        self.progress("Waiting for heading %.02f with accuracy %.02f" %
-                      (heading, accuracy))
-        last_print_time = 0
-        while True:
-            now = self.get_sim_time_cached()
-            if now - tstart >= timeout:
-                break
-            m = self.mav.recv_match(type='VFR_HUD', blocking=True)
-            if now - last_print_time > 1:
-                self.progress("Heading %d (want %f +- %f)" %
-                              (m.heading, heading, accuracy))
-                last_print_time = now
-            if math.fabs(m.heading - heading) <= accuracy:
-                self.progress("Attained heading %d" % heading)
-                return True
-        raise WaitHeadingTimeout("Failed to attain heading %d" % heading)
+        def get_heading_wrapped(target, timeout2):
+            msg = self.mav.recv_match(type='VFR_HUD', blocking=True, timeout=timeout2)
```

Comment:
You're just dropping out the end of these `get` functions.  That's a little untidy.

Reply:
I remove the target leftover